Wire sync and registry jobs

JulzDiverse opened the pull request:
This PR wires the sync and registry job and, thus, removes the hardcoded `busybox` from the `dropletToImageUri` function. 

This PR contains the following changes:

**implement `dropletToImageUrl()`**

Based on the Desired App, this code:

- downloads the droplet, 
- sends the droplet to the registry which converts and adds the droplet to the local blob (the droplets are saved by the droplet hash)
- and generates the droplet URL.

The cfclient, http client, and other required parameters are passed from the `main` function. Therefore the `Convert` function signiture is enhanced. 

**Introduce cfclient**

As the sync loop requires to download droplets from the `cloundcontroller/blobstore` and the internal clound controller api endpoint cannot be resolved, the public [api endpoint](https://apidocs.cloudfoundry.org/280/apps/downloads_the_staged_droplet_for_an_app.html) is used. The `cfclient` simplifies the process to get the bearer token by just providing the cf admin user and password.

The cfclient code reuses the `client.go` (and some more required code) from https://github.com/cloudfoundry-community/go-cfclient and enhances the client code with the `GetDropletByAppGuid` method. 

**Introduce models.go**

 Having all domain types in a `model.go` has the following advantages:

- Root package is for domain types
- Group subpackages by dependency
- Use a shared mock subpackage
- Main package ties together dependencies

More about this approach can be read [here](https://medium.com/@benbjohnson/standard-package-layout-7cdbc8391fc1).

**Enhance sync CLI**

The sync CLI add further parameters to setup the cfclient, which is required to download the droplets. Moreover, the CLI was enhanced to consume a config YAML file instead of passing CLI flags. 

**Update Tests**

The tests were updated according the code changes. 

**Update sink README**

**Add some output in the registry when droplets are added**

**The code was tested on a local Bosh-Lite and a Minikube cluster (also in combination with the cube-release)** ([here](https://github.com/andrew-edgar/cube-release/pull/14) is the PR for the cube-release)

julz commented:
any chance we could tidy the commits up a bit here before we merge? the merge commit at the end makes it a bit hard to see what actually happened and I can't find a single commit where dropletToImage went in that doesn't also include lots of dependency additions. Maybe a commit to introduce the new dependencies and a commit to implement dropletToImage or something?

JulzDiverse commented:
@julz 
I cleaned up the commit history into 4 commits:

- implementation of `dropletToImageUri`
- adding `cfclient`
- updating sync CLI
- small registry updates and vendor updates

JulzDiverse commented on registry/registry.go:
```diff
@@ -41,7 +43,9 @@ func NewHandler(rootfsBlob BlobRef, dropletStore DropletStore, blobs BlobStore)
 
 func Ping(w http.ResponseWriter, r *http.Request) {
-	// 200
+	w.WriteHeader(http.StatusOK)
```
Martini is imported by the `cfclient/client.go` code, not by registry. 
The last thing i did after my changes was to updated vendors. So this commit contains all required dependencies for all my changes.

julz replied:
ah, we probably should've had the dependencies in an earlier commit, then; ideally every commit should compile and pass tests (so that `git bisect` works). No big deal, will just squash the commits before merging.

JulzDiverse replied:
i think i did that in my last update. I updated dependencies on each commit. It should be better after i reopened the PR.

JulzDiverse commented:
@julz 
I rewrote the error handling for the Convert function as discussed. It panics inside `dropletToImageUrl` and `ConvergeOnce` recovers and logs the error. Furthermore:

- All lager messages now match json keys. 
- I updated the dependencies per commit.

julz commented:
cool, in retrospect I think letting the panic escape from Convert might be a bit bad (since it's an exported function), not a big deal though we can always tidy that up later. Gonna merge.
